zproject/backends.py

```diff
@@ -1901,6 +1933,21 @@ def sync_groups(
     logger.debug("Finished group sync for user %s", user_id)
 
 
+def ensure_missing_groups(missing_group_names: set[str], realm: Realm) -> list[NamedUserGroup]:
+    system_groups_name_dict = get_role_based_system_groups_dict(realm)
+    group_owners = system_groups_name_dict[SystemGroups.OWNERS].usergroup_ptr
+    group_settings_map = dict(
+        can_add_members_group=group_owners,
+        can_manage_group=group_owners,
+    )
+    return [
+        create_user_group_in_database(
+            group_name, [], realm, acting_user=None, group_settings_map=group_settings_map
```

Comment:
What's our story about race conditions here?

We have constraints on group naming -- see `check_user_group_name`.  We aren't validating those here, which means is the SAML source asks us, we will make a NamedUserGroup named e.g. `role:owners` ... which seems Bad.  We need to apply those checks.

And if the remote system wants to make a group name which we don't allow, should that 500 the login, or bravely soldier on with most of the group memberships in place?

Will admins find it unexpected that the new group will only be _partially_ sync'd, in that it will not include members that joined Zulip before this deploy (i.e. it's not retroactive)?

Reply:
I think that reasoning makes sense.

Reply:
ok, wrapped `ensure_missing_groups` in `@transaction.atomic` to address the race concern.